### PR TITLE
Formula filling

### PR DESCRIPTION
**Pull request by mmcdermott**
This adds a small formula fill nub onto the corner of fillable flow grid cells and allows you to drag to fill that cell over a region. This will intelligently fill functions and static objects across a the desired row or column.

Has one noteworthy bug which is that the fill nub renders above the metric card reasoning popout.

### CODE REVIEW

File: src/components/lib/FlowGrid/FlowGrid.css [hunk 1]
```diff
@@ -119,7 +119,44 @@ svg.edge  path.basic-arrow.RED {
...
+.FlowGridCell .FillToken {
+   cursor: crosshair;
+   height: 100%;
+   width: 100%;
+   border-radius: 0.1em;
+   background: rgb(90, 141, 177);
+ }
+
+.FlowGridCell .FillToken--outer:hover .FillToken {
+  transition: border .05s;
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.css]: remove

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 2]
```diff
@@ -82,13 +87,30 @@ export default class FlowGrid extends Component{
     return !sameLocation
   }
 
+  newFillRegion(end) {
+    const {fillRegion: {start}} = this.state
+    if (Math.abs(end.row - start.row) > Math.abs(end.column - start.column)) {
+      return {start, end: {row: end.row, column: start.column}}
+    } else {
+      return {start, end: {row: start.row, column: end.column}}
+    }
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: can the above code be cleaned up a bit? A bit tricky to understand.

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 3]
```diff
@@ -82,13 +87,30 @@ export default class FlowGrid extends Component{
...
+  }
+
   _handleCellMouseEnter(location, e) {
     window.recorder.recordNamedEvent("FlowGrid set hover state")
-    if (this.state.leftDown && this._mouseMoved(e)) {
-      this.setState({hover: {row: -1, column: -1}})
-      this._handleEndRangeSelect(location)
-    } else {
+    if (!(this._mouseMoved(e) && (this.state.tracingFillRegion || this.state.dragSelecting))) {
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: This should be split up or something.  Quite terse for one line.

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 4]
```diff
@@ -82,13 +87,30 @@ export default class FlowGrid extends Component{
...
       this.setState({hover: location})
+      return
+    }
+    const hover = {row: -1, column: -1}
+
+    if (this.state.tracingFillRegion) {
+      window.recorder.recordNamedEvent("FlowGrid set fillRegion state")
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: single quotes

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 5]
```diff
@@ -82,13 +87,30 @@ export default class FlowGrid extends Component{
...
+    const {fillRegion: {start}} = this.state
+    if (Math.abs(end.row - start.row) > Math.abs(end.column - start.column)) {
+      return {start, end: {row: end.row, column: start.column}}
+    } else {
+      return {start, end: {row: start.row, column: end.column}}
+    }
+  }
+
   _handleCellMouseEnter(location, e) {
     window.recorder.recordNamedEvent("FlowGrid set hover state")
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: single quotes

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 6]
```diff
@@ -198,19 +220,40 @@ export default class FlowGrid extends Component{
...
+
+  onCellMouseUp(location) {
+    if (!this.state.tracingFillRegion) { return }
+    window.recorder.recordNamedEvent("FlowGrid set fillRegion state")
+    this.props.onFillRegion(this.state.fillRegion)
+    this._handleEndRangeSelect(this.state.fillRegion.end)
+    this.setState({tracingFillRegion: false, fillRegion: {}})
+  }
+
+  // TODO(matthew): Look into necessity of 'inSelectedRegion' passed to cell below.
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: Should this TODO be here for the PR?

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 7]
```diff
@@ -198,19 +220,40 @@ export default class FlowGrid extends Component{
...
   _cell(location) {
-    const item = this.props.items.find(i => isAtLocation(i.location, location));
+    const item = this.props.items.find(i => isAtLocation(i.location, location))
+    const {selectedCell, selectedRegion} = this.props
+    const inSelectedCell = isAtLocation(selectedCell, location)
+    const selectedRegionNontrivial = selectedRegion.length === 2 && !isAtLocation(selectedRegion[0], selectedRegion[1])
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: what does nontrivial mean?  better name maybe?

File: src/components/lib/FlowGrid/FlowGrid.jsx [hunk 8]
```diff
@@ -198,19 +220,40 @@ export default class FlowGrid extends Component{
...
+    const hasNonEmptyItem = !!item && !this.props.isItemEmpty(item.key)
+    const showFillToken = inSelectedCell && hasNonEmptyItem && !this.state.dragSelecting && !selectedRegionNontrivial
```
Comment (OAGr) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: This kind of seems like it should be its own function.  Realize that would require some more work though.

File: src/components/lib/FlowGrid/cell.jsx [hunk 9]
```diff
@@ -151,14 +154,28 @@ export default class Cell extends Component {
     return classes
   }
 
+  onFillTargetMouseDown(e) {
+    if (e.button === 0) {
+      this.props.onFillTargetMouseDown()
+      e.stopPropagation()
+      e.preventDefault()
```
Comment (OAGr) [inline src/components/lib/FlowGrid/cell.jsx]: Are you sure that both preventDefault() and stopPropagation() are necessary?

File: src/components/lib/FlowGrid/filled-cell.js [hunk 10]
```diff
@@ -1,7 +1,74 @@
...
 import {PTLocation} from 'lib/locationUtils'
+import { getEmptyImage } from 'react-dnd-html5-backend';
+
+var layerStyles = {
+  position: 'fixed',
+  pointerEvents: 'none',
+  zIndex: 10000,
+  left: 0,
+  top: 0,
+  width: '193px',
```
Comment (OAGr) [inline src/components/lib/FlowGrid/filled-cell.js]: This is quite a hack, I need to fix this.

File: src/components/lib/FlowGrid/filled-cell.js [hunk 11]
```diff
@@ -1,7 +1,74 @@
...
+    item: monitor.getItem(),
+    itemType: monitor.getItemType(),
+    currentOffset: monitor.getSourceClientOffset(),
+    isDragging: monitor.isDragging()
+}))
+export default class DragPreview extends Component {
+  renderItem(type, item) {
+    const styles = {
+      transform: 'rotate(-4deg)',
+      WebkitTransform: 'rotate(-4deg)',
```
Comment (OAGr) [inline src/components/lib/FlowGrid/filled-cell.js]: Note: Remove transforms and marginTop

File: src/lib/locationUtils.js [hunk 12]
```diff
@@ -25,6 +25,18 @@ export function isWithinRegion(test, region) {
...
 }
 
+export function getBounds({start, end}) {
+  if (!start || !end) {return []}
+  let leftX, topY, rightX, bottomY
+  leftX = Math.min(start.row, end.row)
+  topY = Math.max(start.column, end.column)
+  rightX = Math.max(start.row, end.row)
+  bottomY = Math.min(start.column, end.column)
+  return [{row: leftX, column: bottomY}, {row: rightX, column: topY}]
```
Comment (OAGr) [inline src/lib/locationUtils.js]: It's not obvious what these are from the location.  What about this being a hash, {topLeft:{row,column}, bottomRight:...}

File: src/lib/locationUtils.js [hunk 13]
```diff
@@ -25,6 +25,18 @@ export function isWithinRegion(test, region) {
   )
 }
 
+export function getBounds({start, end}) {
+  if (!start || !end) {return []}
```
Comment (OAGr) [inline src/lib/locationUtils.js]: If there's either a start or and end, would we want this to return the bounds being that one element?

File: src/modules/fill_region/actions.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+import {runSimulations, deleteSimulations} from 'gModules/simulations/actions'
+import {registerGraphChange} from 'gModules/spaces/actions'
+
+import e from 'gEngine/engine'
+
+import {isAtLocation, isWithinRegion, getBounds, move, translate} from 'lib/locationUtils'
+
+const DYNAMIC_FILL_TYPE = 'FUNCTION'
+
+// TODO(matthew): Dry up code with this and copy and undo.
```
Comment (OAGr) [inline src/modules/fill_region/actions.js]: is this for after the PR?

File: src/components/lib/FlowGrid/FlowGrid.jsx (hunk 6 above)
Comment (mmcdermott) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: I want to look into it later, so I'm leaving the TODO there for future reference.

File: src/components/lib/FlowGrid/FlowGrid.jsx (hunk 8 above)
Comment (mmcdermott) [inline src/components/lib/FlowGrid/FlowGrid.jsx]: I'm going to leave this as is for now; seems like making it its own function wouldn't actually absolve much complexity, just move it.

File: src/lib/locationUtils.js (hunk 12 above)
Comment (mmcdermott) [inline src/lib/locationUtils.js]: I think that'd be a good change but I'd want to make that across the suite of location code, not just for the auto fill region.

File: src/lib/locationUtils.js (hunk 13 above)
Comment (mmcdermott) [inline src/lib/locationUtils.js]: Possibly, but we don't have that need right now. I think I'd rather just have the api be that you pass in both bounds, and the function computes as such. to get a one element array you can always just pass start, start.

File: src/modules/fill_region/actions.js (hunk 14 above)
Comment (mmcdermott) [inline src/modules/fill_region/actions.js]: Yes, across the action suite.

**Pull request merged**